### ISSUE TITLE
Speedup orchestration

### ISSUE THREAD

**Issue by agaoglu** [rook/rook#4642]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Rook orchestration for ceph should be able to complete faster.

**What is use case behind this feature:**
For an already running or partially running cluster, an orchestration run spends too much time waiting for osd deployment updates. There always is a need for a second (and third...) orchestration run for some reason or another. When that happens orchestrator starts/restarts all prepare jobs and according to their results it tries to update osd deployments *if needed*. Shown in an example log:

```
2019-12-12 19:37:05.436036 I | op-osd: deployment for osd 38 already exists. updating if needed
2019-12-12 19:37:05.440446 I | op-mon: this is not an upgrade, not performing upgrade checks
2019-12-12 19:37:05.440472 I | op-k8sutil: updating deployment rook-ceph-osd-38
2019-12-12 19:37:07.458671 I | op-k8sutil: finished waiting for updated deployment rook-ceph-osd-38
```

The problem is this 2 second sleep between the last line and the previous one. For large clusters these add up to a lot of time since all this is being run sequentially.

Easiest solution would be sleeping for a shorter duration [here](https://github.com/rook/rook/blob/master/pkg/operator/k8sutil/deployment.go#L106) maybe letting it retry a few times.

Sound solution would be watching for the changes in the deployment but I think it'll be bit more involved.

But actually I guess it'd be better if we were able to detect if the deployment is **really needed** to be changed and not even try to update, since in most cases, it doesn't seem to be.

**Environment**:
A cluster with some hundreds of osds for now. Currently the total duration is around 20-25 minutes, needless to say a huge part of it is waiting for a deployment updates. Larger clusters are expected to wait hours.

**Comment by leseb** [comment]
Yes, the operator waits up to 30 times every 2 seconds. We can change this to be up to 60 times every second.

@travisn any idea why 2 seconds?
@agaoglu does that sound reasonable to you?

Also, we thought of being able to tell if there is an upcoming change of not. However, if there is nothing to change, the update will be a no-op so very fast.

**Comment by travisn** [comment]
There are two issues here:
1. We need to understand why the orchestration is triggered 2-3 times. There are several expected reasons for this behavior such as if a new node was detected or a new disk was added. 
   - If watching for new disks is not expected or needed, you could disable the [hotplug](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L155) setting. If it is something else, the operator log should show what triggered it. 
   - If it's not obvious, we may need to enable [debug logging](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L116) to take a look.
1. If the OSD pod specs didn't actually change, the operator shouldn't wait any time for the pod to be updated. We're waiting for nothing to happen.

**Comment by agaoglu** [comment]
@leseb 1 second would mean half total time like 10 minutes instead of 20. Or 2 hours instead of 4. It might be OK when necessary, like during an upgrade, since we'd wait more than a few seconds for each osd to restart, but I think we could do better if we know we won't change anything.

@travisn 1. In my case, most re-reruns are because operator times out waiting for prepare jobs. It takes a bit more than an hour for a node but operator times out after 20 minutes, restarts orchestration, then times out again... This happens only on initialization, so it takes considerably more time to reach a running cluster. It'll be good if we can increase that 20 minute timeout but that's something else.

Adding new disks is pretty common, replacing disks is more like it. Adding nodes not that uncommon because the only reasonable thing we could come up with is to delete any failed nodes and let them add themselves back after they'd been fixed. Latter scenario is not that big a problem, because no deployments need changing, ceph status won't change even after a full orchestration. For the former, we generally want a replaced disk in the cluster, as soon as we can, so it'd be great if we didn't need to wait for all deployments. I alsa saw some spontaneous re-runs but I'm not really sure TBH, but again it's generally not a big problem if we're not waiting for something to happen ceph-wise.

2. Totally agree.

**Comment by agaoglu** [comment]
Now that #4442 is released in 1.3 I'm retrying to see the updated behavior and it's a bit interesting.

```
2020-04-16 12:10:01.979446 I | cephclient: getting or creating ceph auth key "osd.552"
2020-04-16 12:10:02.770107 I | op-osd: deployment for osd 552 already exists. updating if needed
2020-04-16 12:10:03.170659 I | op-k8sutil: deployment "rook-ceph-osd-552" did not change, nothing to update
2020-04-16 12:10:03.170702 I | op-osd: started deployment for osd 552
2020-04-16 12:10:03.170719 I | cephclient: getting or creating ceph auth key "osd.144"
2020-04-16 12:10:03.969582 I | op-osd: deployment for osd 144 already exists. updating if needed
2020-04-16 12:10:04.370707 I | op-k8sutil: deployment "rook-ceph-osd-144" did not change, nothing to update
2020-04-16 12:10:04.370734 I | op-osd: started deployment for osd 144
2020-04-16 12:10:04.370746 I | cephclient: getting or creating ceph auth key "osd.259"
2020-04-16 12:10:05.169520 I | op-osd: deployment for osd 259 already exists. updating if needed
2020-04-16 12:10:05.570247 I | op-k8sutil: deployment "rook-ceph-osd-259" did not change, nothing to update
2020-04-16 12:10:05.570274 I | op-osd: started deployment for osd 259
2020-04-16 12:10:05.570286 I | cephclient: getting or creating ceph auth key "osd.276"
2020-04-16 12:10:06.372512 I | op-osd: deployment for osd 276 already exists. updating if needed
2020-04-16 12:10:06.770676 I | op-k8sutil: deployment "rook-ceph-osd-276" did not change, nothing to update
2020-04-16 12:10:06.770700 I | op-osd: started deployment for osd 276
```

That 2 second delay is gone, but it seems we're now waiting for a ceph auth key retrieval for about 0.8 seconds and another 0.4 seconds to check for differences. Not really sure if that auth key step can be avoided as it wasn't there in 1.2. I'll be checking it a bit more but if there are any ideas in the meantime, I'll be grateful.

**Comment by agaoglu** [comment]
OK that's definitely not about creating a ceph key. It seems it was there all along and 1.3 only started logging it. I don't think there's much to improve here at this point. These are all designed for initial provisioning while I'm trying to optimize something else. Thanks.
